src/writer.rs

```diff
@@ -804,14 +845,84 @@ mod tests {
         let too_large_mem_reserve: Vec<FdtReserveEntry> = (0..overflow_size)
-            .map(|i| FdtReserveEntry {
-                address: u64::from(i) * 2,
-                size: 1,
-            })
+            .map(|i| FdtReserveEntry::new(u64::from(i) * 2, 1).unwrap())
             .collect();
         let mut fdt = FdtWriter::new(&too_large_mem_reserve).unwrap();
         let root_node = fdt.begin_node("").unwrap();
         fdt.end_node(root_node).unwrap();
         assert_eq!(fdt.finish().unwrap_err(), Error::TotalSizeTooLarge);
     }
+
+    #[test]
+    fn test_invalid_mem_reservations() {
+        // Test that we cannot create an invalid FDT reserve entry where the
+        // end address of the region would not fit in an u64.
+        assert_eq!(
+            FdtReserveEntry::new(0x1, u64::MAX).unwrap_err(),
+            Error::InvalidMemoryReservation
+        );
+
+        // Test that we cannot have a memory reservation with size 0.
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 0).unwrap_err(),
+            Error::InvalidMemoryReservation
+        );
+    }
+
+    #[test]
+    fn test_cmp_mem_reservations() {
+        // Test that just the address is taken into consideration when comparing to `FdtReserveEntry`.
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 10)
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x1, 11).unwrap()),
+            Ordering::Equal
+        );
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 10)
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x2, 10).unwrap()),
+            Ordering::Less
+        );
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 10)
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x0, 10).unwrap()),
+            Ordering::Greater
+        );
+    }
+
+    #[test]
+    fn test_overlapping_mem_reservations() {
+        // Check that regions that overlap return an error on new.
+        // Check overlapping by one.
+        let overlapping = [
+            FdtReserveEntry::new(0x3, 1).unwrap(), // this overlaps with
+            FdtReserveEntry::new(0x0, 1).unwrap(),
+            FdtReserveEntry::new(0x2, 2).unwrap(), // this one.
+        ];
+        let fdt = FdtWriter::new(&overlapping);
+        assert_eq!(fdt.unwrap_err(), Error::OverlappingMemoryReservations);
+
+        // Check a larger overlap.
+        let overlapping = [
+            FdtReserveEntry::new(0x100, 100).unwrap(),
+            FdtReserveEntry::new(0x50, 300).unwrap(),
+        ];
+        let fdt = FdtWriter::new(&overlapping);
+        assert_eq!(fdt.unwrap_err(), Error::OverlappingMemoryReservations);
+    }
+
+    #[test]
+    fn test_off_by_one_mem_rsv() {
+        // This test is for making sure we do not introduce off by one errors
+        // in the memory reservations checks.
+        let non_overlapping = [
+            FdtReserveEntry::new(0x0, 1).unwrap(), // this overlaps with
```

Comment:
Maybe a copy-pasted comment from the other test? This one doesn't actually overlap.

Reply:
Yes :( true